### PR TITLE
add support of OAuth 2.0/OIDC 'code with PKCE' flow (back-end)

### PR DESCRIPTION
**Pull request by laminelam**
https://issues.apache.org/jira/browse/SOLR-16897

This is the "back-end" part of '[code with PKCE](https://github.com/apache/solr/pull/1791)' contribution. 

The back-end code is mainly for configuration. This is where the different options are configured.

This PR adds _tokenEndpoint_ and _authorizationFlow_ options.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by risdenk — APPROVED** [review]

File: solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTIssuerConfig.java [hunk 1]
```diff
@@ -315,6 +327,24 @@ public JWTIssuerConfig setAuthorizationEndpoint(String authorizationEndpoint) {
...
+    this.tokenEndpoint = tokenEndpoint;
+    return this;
+  }
+
+  public String getAuthorizationFlow() {
+    return authorizationFlow;
+  }
+
+  public JWTIssuerConfig setAuthorizationFlow(String authorizationFlow) {
+    this.authorizationFlow = authorizationFlow;
```
Comment (janhoy) [inline solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTIssuerConfig.java:344]: Should we not validate the flow value early and fail if not either of the two allowed values? I'm thinking we can also let this default to `implicit` here in the backend instead of in the frontend?

**Review by janhoy — COMMENTED** [review]
Looks overall good.

Please do a stab on RefGuide docs. The ref guide should encourage the use of code flow and warn that implicit flow support is deprecated.

Perhaps we can also print a WARN log in backend if `implicit` is in use (either explicitly or implicitly) so that users upgrading to 9.4 without reading the release notes will at least get a warn log that they should switch.

File: solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java [hunk 2]
```diff
@@ -847,9 +851,11 @@ protected String generateAuthDataHeader() {
     Map<String, Object> data = new HashMap<>();
     data.put(
         JWTIssuerConfig.PARAM_AUTHORIZATION_ENDPOINT, primaryIssuer.getAuthorizationEndpoint());
+    data.put(JWTIssuerConfig.PARAM_TOKEN_ENDPOINT, primaryIssuer.getTokenEndpoint());
```
Comment (janhoy) [inline solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java:854]: In the `X-Solr-AuthData` the keys should be "snake_case", so this should be `token_endpoint`, not `tokenEndpoint`. The frontend part will probably need to adapt to this as well.
```suggestion
    data.put("token_endpoint", primaryIssuer.getTokenEndpoint());
```

File: solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java [hunk 3]
```diff
@@ -847,9 +851,11 @@ protected String generateAuthDataHeader() {
     Map<String, Object> data = new HashMap<>();
     data.put(
         JWTIssuerConfig.PARAM_AUTHORIZATION_ENDPOINT, primaryIssuer.getAuthorizationEndpoint());
```
Comment (janhoy) [inline solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java:853]:
```suggestion
        "authorization_endpoint", primaryIssuer.getAuthorizationEndpoint());
```
See comment in line below

Comment (janhoy) [reply solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java:853]: For this line I realize that it's been like this since 9.0, so let's not change that in this PR...

File: solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTIssuerConfig.java (hunk 1 above)
Comment (laminelam) [reply solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTIssuerConfig.java]: I've added a validation test to _setAuthorizationFlow_. The reason I am not doing it earlier is:
- I am following the same logic/style than the other methods (like _setJwksUrl_)
- Adding it earlier would lead to too much verbose code for such basic test.

I have also set _implict_ as default. Will change the front code too.

File: solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java (hunk 3 above)
Comment (laminelam) [reply solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java:853]: Okay, I am leaving as it is both _PARAM_AUTHORIZATION_ENDPOINT_ and _PARAM_TOKEN_ENDPOINT_  for now, is that what you meant?

**Comment by laminelam** [comment]
> Looks overall good.
> 
> Please do a stab on RefGuide docs. The ref guide should encourage the use of code flow and warn that implicit flow support is deprecated.
> 
> Perhaps we can also print a WARN log in backend if `implicit` is in use (either explicitly or implicitly) so that users upgrading to 9.4 without reading the release notes will at least get a warn log that they should switch.

Thank you @janhoy for the review.
I am planning to add a new entry to the ref guide.
BTW, I have added a WARN message.

File: solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java (hunk 3 above)
Comment (janhoy) [reply solr/modules/jwt-auth/src/java/org/apache/solr/security/jwt/JWTAuthPlugin.java:853]: Yea, let’s re-visit this later..

**Comment by janhoy** [comment]
> BTW, I have added a WARN message.

Thanks, the validation and warn log looks good 👍

Will do a final review when refguide docs are in. Also remember CHANGES.txt.

**Review by janhoy — CHANGES_REQUESTED** [review]
* Add refguide docs
* CHANGES entry
* Constant naming
* Backend part of CSP header

**Comment by janhoy** [comment]
Eager to get this into 9.4. Let me know if we can help finalizing these two PRs

**Comment by laminelam** [comment]
> Eager to get this into 9.4. Let me know if we can help finalizing these two PRs
@janhoy 
I should push a new commit sometime today. I have changed little bit your CSP patch, I just need to add a test class for LoadAdminUiServlet.

**Comment by janhoy** [comment]
> > Eager to get this into 9.4. Let me know if we can help finalizing these two PRs
> 
> @janhoy I should push a new commit sometime today. I have changed little bit your CSP patch, I just need to add a test class for LoadAdminUiServlet.

Cool. In my patch i completely disabled CSP response for all other endpoints than the `AdminUIServlet`. I cannot see a reason to return that header for other endpoints. The only I can think of is if you have a solr package that exposes another UI on some path, it could be nice to protect that UI with CSP rules, so if it would be possible to have some fallback CSP header for all other requests perhaps?

File: solr/solr-ref-guide/modules/deployment-guide/pages/jwt-authentication-plugin.adoc [hunk 4]
```diff
@@ -85,6 +85,8 @@ jwk                  ; As an alternative to `jwksUrl` you may provide a static J
 iss                  ; Unique issuer id as configured on the IdP. Incoming tokens must have a matching `iss` claim. Also used to resolve issuer when multiple issuers configured.      ; Auto configured if `wellKnownUrl` is provided
 aud                  ; Validates that the `aud` (audience) claim equals this string      ; Uses `clientId` if configured
 authorizationEndpoint; The URL for the Id Provider's authorization endpoint ; Auto configured if `wellKnownUrl` is provided
+tokenEndpoint; The URL for the Id Provider's token endpoint ; Auto configured if `wellKnownUrl` is provided
+authorizationFlow; Specifies the OAuth 2.0 flow to be used. Supported flows are 'implicit_flow' and 'code_pkce' (for authorization code with 'Proof Key for Code Exchange'). Note: 'implicit_flow' is deprecated and it is highly recommended to use 'code_pkce' instead. ; implicit_flow
```
Comment (janhoy) [inline solr/solr-ref-guide/modules/deployment-guide/pages/jwt-authentication-plugin.adoc:89]: `implicit_flow` --> `implicit`?

File: solr/server/etc/jetty.xml [hunk 5]
```diff
@@ -99,6 +100,7 @@
           </New>
         </Arg>
       </Call>
```
Comment (janhoy) [inline solr/server/etc/jetty.xml:102]: Do we need a fallback CSP header for all API endpoints?

File: solr/solr-ref-guide/modules/deployment-guide/pages/jwt-authentication-plugin.adoc (hunk 4 above)
Comment (laminelam) [reply solr/solr-ref-guide/modules/deployment-guide/pages/jwt-authentication-plugin.adoc]: > `implicit_flow` --> `implicit`?

Good catch.
Fixed :-)

**Comment by laminelam** [comment]
> > > Eager to get this into 9.4. Let me know if we can help finalizing these two PRs
> > 
> > 
> > @janhoy I should push a new commit sometime today. I have changed little bit your CSP patch, I just need to add a test class for LoadAdminUiServlet.
> 
> Cool. In my patch i completely disabled CSP response for all other endpoints than the `AdminUIServlet`. I cannot see a reason to return that header for other endpoints. The only I can think of is if you have a solr package that exposes another UI on some path, it could be nice to protect that UI with CSP rules, so if it would be possible to have some fallback CSP header for all other requests perhaps?

Yes I was thinking the same thing, we don't need this header for non browser targeted servlets (API endpoint, etc), so it will be just ignored by other endpoints. Now, I am not sure of a scenario where we would need a different UI servlet exposed on a different path. Let me think about it and get back to you hopefully by tomorrow.

**Comment by laminelam** [comment]
> * Add refguide docs
> * CHANGES entry
> * Constant naming
> * Backend part of CSP header

Will the CHANGEs entry.

Here are the main changes to the patch:
- In line with the security best practice of granting minimal necessary permissions, adjusted the _registerTokenEndpointForCsp_() method to incorporate the precise URL for the /token endpoint rather than a more general host:port based URL.

- Thought providing the capability of passing the URLs as a comma-separated list is less confusing for the end-user/operator compared to using a prefix-based system property name. This approach also enables more concise implementation.

File: solr/server/etc/jetty.xml (hunk 5 above)
Comment (janhoy) [reply solr/server/etc/jetty.xml:102]: @rmuir You added the CSP stuff to `jetty.xml`. Do you see a reason to keep sending the header for any and all API requests to Solr that are not from a browser? I don't.

@chatman The only other UI I can think of being served by Solr backend are those that are part of a package installed by package manager. Those UIs are 100% stand alone, can we also let it be the UI author's responsibility to secure it with proper headers? I think so.

Comment (janhoy) [reply solr/server/etc/jetty.xml:102]: If there's an easy way to provide a "fallback" if the header is not set, we could do that to avoid this discussion. I'm aware of Jetty's generic response rewrite plugins. We also have our `SolrDispatchFilter` but I find no single code point where one could manipulate response header before response is sent.

Comment (rmuir) [reply solr/server/etc/jetty.xml:102]: > @rmuir You added the CSP stuff to `jetty.xml`. Do you see a reason to keep sending the header for any and all API requests to Solr that are not from a browser? I don't.

I set these headers across the board for all responses, pure API servers too. For something doing OAuth authorization code flow I would definitely set it.  I don't understand "not from a browser" as this flow usually involves one?

**Comment by epugh** [comment]
We do have two packages "in the wild" that come with a UI, the YASA package https://github.com/yasa-org/yasa and Splainer, https://github.com/o19s/splainer/tree/main/solr-splainer-package.    Will these packages need to be updated?   

If we want to make this change, and that ripples down to packages, then we should do it now before more Packages are developed ;-).

We do have a test (commented out) that validates deploying Packages that might be useful: https://github.com/apache/solr/blob/main/solr/packaging/test/test_packages.bats#L69

File: solr/server/etc/jetty.xml (hunk 5 above)
Comment (janhoy) [reply solr/server/etc/jetty.xml]: Ok, turned out to be much simpler. Just leaving the `HeaderPatternRule` in `jetty.xml` worked fine. The `LoadAdminUiServlet` simply overwrite the header, and everything is just as it has been, no extra code. I pushed the fix, we can close this concern.

**Comment by janhoy** [comment]
> We do have two packages "in the wild" that come with a UI

As stated anove, we never really needed to disable the old CSP header config in `jetty.xml`. Turns out the `LoadAdminUiServlet` simply replaces the header, which is nice.

File: solr/core/src/java/org/apache/solr/servlet/LoadAdminUiServlet.java [hunk 6]
```diff
@@ -60,20 +66,49 @@ public void doGet(HttpServletRequest _request, HttpServletResponse _response) th
...
+
+  /**
+   * Retrieves the specification version of the SolrCore package.
+   *
+   * <p>This method is primarily introduced to facilitate unit testing
+   *
+   * @return The specification version of the SolrCore class's package or Unknown if it's
+   *     unavailable.
+   */
+  protected String getSolrCorePackageSpecVersion() {
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/servlet/LoadAdminUiServlet.java:99]: Precommit checks fail due to this. Change to private:
```suggestion
  private String getSolrCorePackageSpecVersion() {
```

Comment (laminelam) [reply solr/core/src/java/org/apache/solr/servlet/LoadAdminUiServlet.java]: done

File: solr/CHANGES.txt [hunk 7]
```diff
@@ -147,6 +149,8 @@ Bug Fixes
 * SOLR-16916: Use of the JSON Query DSL should ignore the defType parameter
   (Christina Chortaria, Max Kadel, Ryan Laddusaw, Jane Sandberg, David Smiley)
 
+* SOLR-16900: Add/fix wrong/missing "solr.home" to zkcli scripts  (Lamine Idjeraoui. Review by Eric Pugh & Jason Gerlowski)
```
Comment (janhoy) [inline solr/CHANGES.txt:152]: Unrelated edit?

File: solr/CHANGES.txt [hunk 8]
```diff
@@ -107,6 +107,8 @@ Improvements
 
 * SOLR-16927: Allow SolrClientCache clients to use Jetty HTTP2 clients (Alex Deparvu, David Smiley)
 
+* SOLR-16897: Add support of OAuth 2.0/OIDC 'code with PKCE' flow - back-end  (Lamine Idjeraoui. Review and contribution by janhoy. Review by Kevin Risden)
```
Comment (janhoy) [inline solr/CHANGES.txt:110]: Simply list the contributors, no need to qualify what each one contributed.

Github will set author and add `Co-authored-by: John Doe <john@doe.com>` for code contributors in final merge message.

Also, no ned for two CHANGES entris even if you split it in two PRs with two commits. Suggest to add both JIRA numbers in same line:

```suggestion
* SOLR-16896, SOLR-16897: Add support of OAuth 2.0/OIDC 'code with PKCE' flow (Lamine Idjeraoui, janhoy, Kevin Risden)
```

File: solr/CHANGES.txt (hunk 7 above)
Comment (laminelam) [reply solr/CHANGES.txt:152]: Yes not related to this PR.
But I forgot to add a change entry of another already merged PR, so I added it here to avoid a new PR just to add a change entry.
If you think it's not good idea never mind we can remove it.

Comment (janhoy) [reply solr/CHANGES.txt:152]: Yea, remove it from this PR, and do a separate commit (and PR) for the missing changes entry, with `SOLR-16900: ..` in the commit message - that will make it ping the JIRA issue to keep track of the commit history.

Comment (laminelam) [reply solr/CHANGES.txt:152]: done :)

**Review by janhoy — APPROVED** [review]
I fixed the merge conflict and CHANGES entry. Think this is good to go now. Great work!

**Comment by janhoy** [comment]
I'm going to force-push a squashed git-history in order for full test suite to run. Will preserve you as commit author..
